CallSiteValidator doesn't cache results of the walk

christiaanderidder opened the pull request:
See https://github.com/dotnet/runtime/issues/96253

Currently to make sure no scoped dependencies are injected into singleton dependencies, the entire tree of a call site is walked for each dependency registered in the DI container. This happens even when that tree has been walked before when validating another call site.

By making a small adjustment to the existing code, we cache each walked call site tree instead of just the trees containing a scoped dependency in one of the branches.

This massively reduces the time it takes to validate the DI container when ValidateOnBuild is used, at the expense of some additional memory needed to store the cached entries.

This solution can most likely be made slightly more efficient by keeping a separate HashSet for just the visited trees instead of inserting `null` values into the existing ConcurrentDictionary, but I opted to keep the solution as close as possible to the original implementation.

christiaanderidder committed:
Cache all services encountered during scope validation to speed up validate on build

christiaanderidder committed:
Fix typo

christiaanderidder commented:
@dotnet-policy-service agree company="ChannelEngine.com B.V."

christiaanderidder commented:
/azp run runtime

steveharter commented on src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteValidator.cs:
```diff
@@ -10,21 +10,15 @@ namespace Microsoft.Extensions.DependencyInjection.ServiceLookup
     {
         // Keys are services being resolved via GetService, values - first scoped service in their call site tree
-        private readonly ConcurrentDictionary<ServiceCacheKey, Type> _scopedServices = new ConcurrentDictionary<ServiceCacheKey, Type>();
+        private readonly ConcurrentDictionary<ServiceCacheKey, Type?> _scopedServices = new ConcurrentDictionary<ServiceCacheKey, Type?>();
```
@benjaminpetit should this be named-service aware? i.e. use `ServiceCallSite` instead of `Type` here?

benjaminpetit replied:
The `ServiceCacheKey` is already named-service aware, so it should be sufficient

steveharter commented:
Should the tests added to https://github.com/dotnet/extensions/pull/2374 be added here? And the benchmark added to dotnet/performance?

christiaanderidder committed:
Add unit tests added in original PR

christiaanderidder commented:
> Should the tests added to [dotnet/extensions#2374](https://github.com/dotnet/extensions/pull/2374) be added here? And the benchmark added to dotnet/performance?

I had a look at the unit tests and added them in. They don't seem to test anything related to this performance fix (as the behavior should remain exactly the same in that regard), but they do seem to cover some untested edge cases. 

Regarding the benchmark, I have added it to the dotnet/performance repository, the PR can be found here: https://github.com/dotnet/performance/pull/3878

christiaanderidder committed:
Merge branch 'main' into scope-validation-performance

steveharter reviewed (approved):
Thanks!

davidfowl commented:
Seems like there might be a regression caused by this fix:

https://github.com/dotnet/runtime/issues/98551
